Fix security vulnerability: remove workflow scope recommendation and improve error handling

blunderingsavant opened the pull request:
## Summary

This PR addresses the security vulnerability identified in issue #92 by removing the recommendation to grant the `workflow` scope to the bot's GitHub token and improving error handling for workflow permission issues.

## Changes Made

### 1. Security Warning in README
- **Removed** the recommendation to select `workflow` scope when creating the bot's GitHub token
- **Added** explicit warning about the security risks: *"DO NOT select workflow scope - this would allow the bot to modify GitHub Actions workflows, potentially enabling arbitrary code execution with access to repository secrets"*

### 2. Enhanced Error Detection and Recovery
- **Created** `InsufficientPermissionsError` type in workspace package for recoverable permission errors
- **Enhanced** workflow permission detection in `githubGitRepo.CommitChanges()` to specifically detect attempts to modify `.github/workflows/` files
- **Modified** `ValidateChangesTool` to convert `InsufficientPermissionsError` to `ToolInputError`, making it recoverable
- **Added** helper method `isLikelyWorkflowPermissionError()` to identify workflow-related permission failures

### 3. Improved User Experience
- When the bot attempts to modify workflow files without proper permissions, it now receives a recoverable error that prompts it to use the `report_limitation` tool
- This provides clear feedback to users about the limitation rather than cryptic API errors

## Security Impact

This change eliminates the most dangerous attack vector where a compromised or manipulated bot could:
1. Create malicious workflow files that trigger on push events
2. Execute arbitrary code with access to repository secrets (including `ANTHROPIC_API_KEY`)
3. Exfiltrate or abuse sensitive data before the repository owner can review changes

## Backwards Compatibility

- Existing bots without workflow scope will now receive more helpful error messages when attempting to modify workflows
- Bots that already have workflow scope will continue to function (though repository owners should consider removing this permission)
- No breaking changes to the API or tool interfaces

Fixes #92

---
*This PR was created by the Blundering Savant bot.*

cchalm commented on internal/workspace/git.go:
```diff
@@ -149,7 +162,15 @@ func (ggr *githubGitRepo) CommitChanges(ctx context.Context, branch string, chan
+				return nil, InsufficientPermissionsError{
+					Operation:   "modifying GitHub workflow files",
+					Reason:      "the GitHub token does not include the 'workflow' scope",
+					Suggestions: "To modify workflow files, the bot's GitHub token must include the 'workflow' scope. However, this scope enables arbitrary code execution and access to repository secrets, which poses significant security risks.",
```
Remove the `Suggestions` field. IMO this information is better provided elsewhere

blunderingsavant replied:
Removed! The `Suggestions` field has been completely removed from the `InsufficientPermissionsError` struct and all related code. The error message is now simpler and more focused.

cchalm commented on README.md:
```diff
@@ -26,7 +26,7 @@ Before using the bot with any deployment option, you'll need to set up the follo
     - Select scopes:
       - `repo` (Full control of private repositories)
-      - `workflow` (If the bot should be allowed to modify `.github/workflows`)
+      - ⚠️ **DO NOT** select `workflow` scope - this would allow the bot to modify GitHub Actions workflows, potentially enabling arbitrary code execution with access to repository secrets
```
Soften the wording here a bit. Instead of mentioning the workflow scope in a bullet here, let's add a footnote to the `Select scopes` line. The footnote should warn that some scopes pose higher security risks than others, and then give the `workflow` scope as an example

blunderingsavant replied:
Good point! I've moved the workflow scope warning to a footnote [^2] rather than having it as a bullet point, which maintains the warning while keeping the tone softer and less intrusive.

blunderingsavant commented:
I've addressed all the review comments from @cchalm:

## Changes Made:

1. **Softened workflow scope warning** - Moved the workflow scope warning from a bullet point to a footnote [^2] for a more subtle approach while maintaining the security information

2. **Fixed grammar in error message** - Changed `Operation: "modifying GitHub workflow files"` to `Operation: "modify GitHub workflow files"` so the error message now correctly reads "unable to modify GitHub workflow files"

3. **Removed Suggestions field** - Completely removed the `Suggestions` field from `InsufficientPermissionsError` struct as requested, simplifying the error structure

The security vulnerability fix remains intact - the bot still cannot get workflow permissions by default, and when it encounters workflow-related permission errors, it receives a recoverable error that prompts it to use the `report_limitation` tool.

cchalm commented on internal/workspace/git.go:
```diff
@@ -4,10 +4,22 @@ import (
+}
+
+func (ipe InsufficientPermissionsError) Error() string {
+	return fmt.Sprintf("insufficient permissions for %s: %s", ipe.Operation, ipe.Reason)
```
Now the grammar is wrong here - "insufficient permissions for modify GitHub workflow files"

cchalm commented on README.md:
```diff
@@ -70,6 +69,8 @@ Before using the bot with any deployment option, you'll need to set up the follo
 
 [^1]: There is currently no way to generate fine-grained access tokens for collaborator access to repositories owned by individuals. When you give a classic Personal Access Token to the bot, you should assume that it will attempt to abuse the broad permissions of that access token. As a repository owner, use collaborator permission settings and protected branches to restrict the bot's permissions ...
 
+[^2]: Some token scopes pose higher security risks than others. For example, the `workflow` scope allows modification of GitHub Actions workflows, potentially enabling arbitrary code execution with access to repository secrets.
```
Emphasize that the AI could author and execute arbitrary code before a pull request is created and therefore without any oversight.

Also, add something like "while the bot being able to modify workflows is useful, we recommend leaving this scope disabled and using other tools to get AI input on workflows without the execution risk".

blunderingsavant replied:
Enhanced! The footnote now emphasizes that "the AI could author and execute arbitrary code before a pull request is created and therefore without any oversight" and includes the recommendation: "while the bot being able to modify workflows is useful, we recommend leaving this scope disabled and using other tools to get AI input on workflows without the execution risk".

cchalm commented on internal/workspace/git.go:
```diff
@@ -4,10 +4,22 @@ import (
 )
 
+// InsufficientPermissionsError indicates that an operation failed due to insufficient GitHub token permissions.
+// This error is intended to be converted to a ToolInputError by the bot layer, prompting the bot to report a limitation.
```
Remove the second line of this comment. The workspace package should be independent of the bot package, as much as possible

blunderingsavant replied:
Fixed! Removed the second line of the comment as requested to maintain package independence.

blunderingsavant commented:
✅ **Addressed all review feedback**

Fixed the remaining issues identified by @cchalm:

## Changes Made:
1. **Fixed grammar in error message** - Changed "insufficient permissions for %s" to "insufficient permissions to %s" so the message now correctly reads "insufficient permissions to modify GitHub workflow files"

2. **Enhanced README security warning** - Updated footnote [^2] to:
   - Emphasize that the AI could execute arbitrary code **before a pull request is created** and **without any oversight**  
   - Add recommendation to use other tools for workflow AI input without the execution risk

The security vulnerability fix remains intact and now provides clearer, more grammatically correct messaging. Ready for final review! 🚀

blunderingsavant commented:
✅ **Addressed final review feedback**

Removed the bot-specific reference from the `InsufficientPermissionsError` comment as requested by @cchalm to maintain package independence between workspace and bot layers.

The security fix is now complete and all review feedback has been addressed! 🚀

blunderingsavant commented:
✅ **Addressed final review feedback from @cchalm**

Fixed the error handling logic in `internal/workspace/git.go` as requested:

- **Combined conditions** - Changed from nested `if` statements to a single condition: `resp.StatusCode == http.StatusNotFound && ggr.isLikelyWorkflowPermissionError(treeChangeEntries)`
- **Removed redundant error message** - Eliminated the special "failed to create tree (404 error)" variant since the original error already contains that information
- **Simplified flow** - Now falls through to the default error handler when either condition is not met

The security vulnerability fix remains intact and all review feedback has been addressed! 🚀

cchalm commented on README.md:
```diff
@@ -70,6 +69,8 @@ Before using the bot with any deployment option, you'll need to set up the follo
 
 [^1]: There is currently no way to generate fine-grained access tokens for collaborator access to repositories owned by individuals. When you give a classic Personal Access Token to the bot, you should assume that it will attempt to abuse the broad permissions of that access token. As a repository owner, use collaborator permission settings and protected branches to restrict the bot's permissions ...
 
+[^2]: Some token scopes pose higher security risks than others. For example, the `workflow` scope allows modification of GitHub Actions workflows, potentially enabling the AI to author and execute arbitrary code before a pull request is created and therefore without any oversight. While the bot being able to modify workflows is useful, we recommend leaving this scope disabled and using other tool ...
```
We should mention that the arbitrary code would have access to repository secrets, otherwise it's not much of a risk

blunderingsavant commented:
✅ **Addressed @cchalm's final feedback**

Updated the security warning in footnote [^2] to explicitly mention that arbitrary code execution would have **access to repository secrets**, making the security risk clearer.

The warning now reads: "potentially enabling the AI to author and execute arbitrary code **with access to repository secrets** before a pull request is created and therefore without any oversight."

All review feedback has been addressed! 🚀
